### PR TITLE
[DOC] Add tutotrial about improving accuracy of quantization with oneDNN

### PR DESCRIPTION
**Pull request by anko-intel**
## Description ##
Added tutorial showing advantage of using INC with MXNet for quantization. It shows that INC can find operators mostly introduced loss of accuracy and eliminate it from quantization. This way partially quantized model achieves accuracy results almost the same as original floating point model, but with about 3 times performance improvement in comparison to optimized floating point model (or 30 times in comparison to not optimized floating point model). 
Example of results measured on ResNet50 v2 model:
image
image

### Essentials ###
- [x] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented

### CODE REVIEW

**Comment by anko-intel** [comment]
@mxnet-bot run ci [unix-cpu]

**Review by DominikaJedynak — APPROVED** [review]
LGTM

File: docs/python_docs/python/tutorials/performance/backend/dnnl/dnnl_quantization_inc.md [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+# to achieve given accuracy lost against floating point calculation
+print(quantizer.strategy.best_qmodel.q_config['quant_cfg'])
+```
+
+#### Results:
+Resnet50 v2 model could be prepared to achieve better performance with various calibration and tuning methods.  
+It is done by 
+(TODO link to resnet_tuning.py) 
+script on a small part of data set to reduce time required for tuning (9 batches). 
+Later saved model are validated on a whole data set by
```
Comment (bartekkuncer) [inline docs/python_docs/python/tutorials/performance/backend/dnnl/dnnl_quantization_inc.md:246]:
```suggestion
Later saved model is validated on a whole data set by 
```
OR
```suggestion
Later saved models are validated on a whole data set by 
```

File: docs/python_docs/python/tutorials/performance/backend/dnnl/dnnl_quantization_inc.md [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+print(quantizer.strategy.best_qmodel.q_config['quant_cfg'])
+```
+
+#### Results:
+Resnet50 v2 model could be prepared to achieve better performance with various calibration and tuning methods.  
+It is done by 
+(TODO link to resnet_tuning.py) 
+script on a small part of data set to reduce time required for tuning (9 batches). 
+Later saved model are validated on a whole data set by 
+(TODO link to resnet_measurment.py)
```
Comment (bartekkuncer) [inline docs/python_docs/python/tutorials/performance/backend/dnnl/dnnl_quantization_inc.md:247]: TODO?

File: docs/python_docs/python/tutorials/performance/backend/dnnl/dnnl_quantization_inc.md [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+| int8 INC basic       | 0.7692 | 0.9331 |  0.1 |  0.1 | 526.47 |   48.68 | 29.8 |
+| int8 INC mse         | 0.7692 | 0.9337 |  **0.1** |  **0.0** | 227.89 |   50.19 | **28.9** |
+
+Environment:  
+- Intel(R) Xeon(R) Platinum 8375C CPU @ 2.90GHz (c6i.16xlarge Amazon EC2 instance)  
+- Ubuntu 20.04.4 LTS (GNU/Linux Ubuntu 20.04.4 LTS 5.15.0-1017-aws ami-0558cee5b20db1f9c)  
+- MXNet 2.0.0b20220823 (commit daac02c7854ffa71bc11fd950c2d6c9ea356b394 ) 
+- INC 1.13.1  
+- scripts above were run as parameter for [run.sh](https://github.com/apache/incubator-mxnet/blob/master/benchmark/python/dnnl/run.sh) 
+script to properly setup parallel computation parameters
```
Comment (bartekkuncer) [inline docs/python_docs/python/tutorials/performance/backend/dnnl/dnnl_quantization_inc.md:268]:
```suggestion
script to properly setup parallel computation parameters.  
```

File: docs/python_docs/python/tutorials/performance/backend/dnnl/dnnl_quantization_inc.md [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+script to properly setup parallel computation parameters  
+
+For this model INC basic and mse strategies found configurations meeting the 1.5% relative accuracy 
+loss criterion. Only the `bayesian` strategy didn't find solution within 500 attempts limit. 
+Although these results may suggest that the `mse` strategy is the best compromise between time spent
+to find the optimized model and final model performance efficiency, different strategies may give 
+better results for specific models and tasks. You can notice that most imported thing done by INC 
+is to find operator which mostly influence loss of accuracy and disable it from quantization if needed. 
+You can see below which operator was excluded by `mse` strategy in last print given by 
+(TODO link to resnet_mse.py)
```
Comment (bartekkuncer) [inline docs/python_docs/python/tutorials/performance/backend/dnnl/dnnl_quantization_inc.md:277]: TODO?
```suggestion
better results for specific models and tasks. You can notice, that the most important thing done by INC
was to find the operator, which had the most significant impact on the loss of accuracy and disable it from quantization if needed. 
You can see below which operator was excluded by `mse` strategy in last print given by 
(TODO link to resnet_mse.py)
```

File: example/quantization_inc/resnet_measurment.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+  print('{:21} Top1 Accuracy: {:.4f} Top5 Accuracy: {:.4f} from {:4} batches in {:8.2f}s'
+        .format(description, top1, top5, count, time_spend))
+
+# Preparing input data
+rgb_mean = (0.485, 0.456, 0.406)
+rgb_std = (0.229, 0.224, 0.225)
+batch_size = 64
+
+start = time.time()
+# Set below proper path to ImageNet data set
```
Comment (bartekkuncer) [inline example/quantization_inc/resnet_measurment.py:46]:
```suggestion
# Set proper path to ImageNet data set below
```

File: example/quantization_inc/resnet_measurment.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+dataset = mx.gluon.data.vision.ImageRecordDataset('../imagenet/rec/val.rec')
+transformer = transforms.Compose([transforms.Resize(256),
+                                  transforms.CenterCrop(224),
+                                  transforms.ToTensor(),
+                                  transforms.Normalize(mean=rgb_mean, std=rgb_std)])
+# Note: As the input data are used many times it is better to prepared data once,
+#       so lazy parameter for transform_first is set to False
```
Comment (bartekkuncer) [inline example/quantization_inc/resnet_measurment.py:53]:
```suggestion
# Note: as the input data is used many times it is better to prepare it once.
#       Therefore, lazy parameter for transform_first is set to False.
```

File: example/quantization_inc/resnet_measurment.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+val_data = mx.gluon.data.DataLoader(
+    dataset.transform_first(transformer, lazy=False), batch_size, shuffle=False)
+val_data.batch_size = batch_size
+time_consumed = time.time() - start
+print("Input data prepared in {:8.2f}s".format(time_consumed))
+
+print("Measure accuracy on whole data set could take a long time. Please wait...")
```
Comment (bartekkuncer) [inline example/quantization_inc/resnet_measurment.py:60]:
```suggestion
print("Measure accuracy on the whole data set could take a long time. Please wait...")
```

File: example/quantization_inc/resnet_mse.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+from mxnet.gluon.model_zoo.vision import resnet50_v2
+from mxnet.gluon.data.vision import transforms
+from mxnet.contrib.quantization import quantize_net
+
+# Preparing input data
+rgb_mean = (0.485, 0.456, 0.406)
+rgb_std = (0.229, 0.224, 0.225)
+batch_size = 64
+num_calib_batches = 9
+# Set below proper path to ImageNet data set
```
Comment (bartekkuncer) [inline example/quantization_inc/resnet_mse.py:28]:
```suggestion
# Set proper path to ImageNet data set below
```

File: example/quantization_inc/resnet_mse.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+dataset = mx.gluon.data.vision.ImageRecordDataset('../imagenet/rec/val.rec')
+# Tuning in INC on whole data set takes too long time so we take only part of the whole data set
+# as representative part of it:
```
Comment (bartekkuncer) [inline example/quantization_inc/resnet_mse.py:31]:
```suggestion
# Tuning with INC on the whole data set takes loads of time. Therefore, we take only a part of the whole data set
# as representative part of it:
```
Maybe `its representation` instead of `representative part of it`?

File: example/quantization_inc/resnet_mse.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+dataset = dataset.take(num_calib_batches * batch_size)
+transformer = transforms.Compose([transforms.Resize(256),
+                                  transforms.CenterCrop(224),
+                                  transforms.ToTensor(),
+                                  transforms.Normalize(mean=rgb_mean, std=rgb_std)])
+# Note: as input data are used many times during tuning it is better to prepared data earlier,
+#       so lazy parameter for transform_first is set to False
```
Comment (bartekkuncer) [inline example/quantization_inc/resnet_mse.py:38]:
```suggestion
# Note: as input data is used many times during tuning it is better to have it prepared earlier.
#       Therefore, lazy parameter for transform_first is set to False.
```

File: example/quantization_inc/resnet_mse.py [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+
+
+from neural_compressor.experimental import Quantization
+quantizer = Quantization("resnet50v2_mse.yaml")
+quantizer.model = net
+quantizer.calib_dataloader = val_data
+quantizer.eval_func = eval_func
+qnet_inc = quantizer.fit().model
+print("INC finished")
+# You can save optimized model for the later use:
```
Comment (bartekkuncer) [inline example/quantization_inc/resnet_mse.py:61]:
```suggestion
# You can save the optimized model for the later use:
```

File: example/quantization_inc/resnet_mse.py [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+qnet_inc.export("__quantized_with_inc")
+# You can see what configurations was applied aby INC and which nodes was excluded from quantization
+# to achieve given accuracy lost against floating point calculation
```
Comment (bartekkuncer) [inline example/quantization_inc/resnet_mse.py:64]:
```suggestion
# You can see which configuration was applied by INC and which nodes were excluded from quantization
# to achieve given accuracy loss against floating point calculation.
```

Comment (bartekkuncer) [reply example/quantization_inc/resnet_mse.py:64]: Maybe `wanted` instead of `given`?

File: example/quantization_inc/resnet_tuning.py [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+
+save_model.count = 0
+
+# Preparing input data
+start = time.time()
+rgb_mean = (0.485, 0.456, 0.406)
+rgb_std = (0.229, 0.224, 0.225)
+batch_size = 64
+num_calib_batches = 9
+# Set below proper path to ImageNet data set
```
Comment (bartekkuncer) [inline example/quantization_inc/resnet_tuning.py:42]:
```suggestion
# Set proper path to ImageNet data set below
```

File: example/quantization_inc/resnet_tuning.py [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+dataset = mx.gluon.data.vision.ImageRecordDataset('../imagenet/rec/val.rec')
+# Tuning in INC on whole data set takes too long time so we take only part of the whole data set
+# as representative part of it:
```
Comment (bartekkuncer) [inline example/quantization_inc/resnet_tuning.py:45]:
```suggestion
# Tuning with INC on the whole data set takes too much time. Therefore, we take only a part of the whole data set
# as representative part of it:
```

File: example/quantization_inc/resnet_tuning.py [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+dataset = dataset.take(num_calib_batches * batch_size)
+transformer = transforms.Compose([transforms.Resize(256),
+                                  transforms.CenterCrop(224),
+                                  transforms.ToTensor(),
+                                  transforms.Normalize(mean=rgb_mean, std=rgb_std)])
+# Note: As the input data are used many times during tuning it is better to prepared data erlier,
+#       so lazy parameter in transform_first is set to False
```
Comment (bartekkuncer) [inline example/quantization_inc/resnet_tuning.py:52]:
```suggestion
# Note: as the input data is used many times during tuning it is better to have it prepared earlier.
#       Therefore, lazy parameter in transform_first is set to False.
```

**Review by bartekkuncer — CHANGES_REQUESTED** [review]

**Comment by bartekkuncer** [comment]
> ## Description
> Added tutorial showing advantage of using INC with MXNet for quantization. It shows that INC can find operators mostly introduced loss of accuracy and eliminate it from quantization. This way partially quantized model achieves accuracy results almost the same as original floating point model, but with about 3 times performance improvement in comparison to optimized floating point model (or 30 times in comparison to not optimized floating point model)
> 
> image image
> 
> ## Checklist
> ### Essentials
> * [x]  PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
> * [x]  Changes are complete (i.e. I finished coding on this PR)
> * [ ]  All changes have test coverage
> * [ ]  Code is well-documented

I assume this graphics depicts the results for resnet50 v2? I believe it would be nice to add information about the workload in the description.

File: docs/python_docs/python/tutorials/performance/backend/dnnl/dnnl_quantization_inc.md [hunk 16]
```diff
@@ -0,0 +1,290 @@
...
+qnet = quantizer.fit().model
+```
+
+Since this model already achieves good accuracy using native quantization (less than 1% accuracy drop), for the given configuration file, INC will end on the first configuration, quantizing all layers using `naive` calibration mode for each. To see the true potential of INC, we need a model which suffers from a larger accuracy drop after quantization.
+
+### Quantizing ResNet50v2
+
+This example shows how to use INC to quantize ResNet50 v2. In this case, the native MXNet quantization introduce a huge accuracy drop (70% using `naive` calibration mode) and INC allows automatically find better solution.
+
+This is the (TODO link to INC configuration file) for this example:
```
[the remark this answers is not in the document]
Comment (anko-intel) [reply docs/python_docs/python/tutorials/performance/backend/dnnl/dnnl_quantization_inc.md:161]: I have to remove links as linkcheck doesn't allow put links to just added files

**Review by bartekkuncer — APPROVED** [review]
LGTM.

**Comment by anko-intel** [comment]
@mxnet-bot run ci [unix-gpu]

**Review by bgawrych — APPROVED** [review]
